### PR TITLE
Optimized updateShaders in forward-renderer

### PR DESCRIPTION
**Pull request by mvaligursky**
Fixes https://github.com/playcanvas/engine/issues/2633

### CODE REVIEW

File: src/scene/forward-renderer.js [hunk 1]
```diff
@@ -2290,59 +2293,55 @@ Object.assign(ForwardRenderer.prototype, {
         // #endif
     },
 
-    updateShaders: function (drawCalls) {
-        // #ifdef PROFILER
-        var time = now();
-        // #endif
+    // funtion returns unique set of materials from drawCalls
```
Comment (willeastcott) [inline src/scene/forward-renderer.js:2296]:
```suggestion
    // function returns unique set of materials from drawCalls
```

File: src/scene/scene.js [hunk 2]
```diff
@@ -140,7 +140,7 @@ function Scene() {
         lastStaticPrepareWriteTime: 0,
         lastStaticPrepareTriAabbTime: 0,
         lastStaticPrepareCombineTime: 0,
-        updateShadersTime: 0
+        updateShadersTime: 0 // deprecated
```
Comment (willeastcott) [inline src/scene/scene.js:143]: Just curious, but why is this now deprecated?

Comment (mvaligursky) [reply src/scene/scene.js:143]: it was only used around the functions I optimized, which take no time now .. so no need to report duration

File: src/scene/forward-renderer.js [hunk 3]
```diff
@@ -2290,59 +2293,55 @@ Object.assign(ForwardRenderer.prototype, {
...
+        return materials;
+    },
+
+    updateShaders: function (drawCalls) {
+
         // Clear material shaders
-        for (i = 0; i < materials.length; i++) {
-            var mat = materials[i];
+        this.getUniqueMaterials(_tempMaterialSet, drawCalls);
+        for (var mat of _tempMaterialSet) {
```
Comment (Maksims) [inline src/scene/forward-renderer.js:2315]: There are two loops here with extra function call. Why not iterate once through draw calls with track of updated materials using a set (as is now), so eliminating for second iteration through set of materials.

File: src/scene/forward-renderer.js [hunk 4]
```diff
@@ -2290,59 +2293,55 @@ Object.assign(ForwardRenderer.prototype, {
...
-                if (mat.updateShader !== Material.prototype.updateShader) {
-                    if (mat.useLighting === false || (mat.emitter && !mat.emitter.lighting)) {
-                        // skip unlit standard and particles materials
-                        continue;
-                    }
-                    mat.clearVariants();
-                    mat.shader = null;
+        // Clear material shaders
+        this.getUniqueMaterials(_tempMaterialSet, drawCalls);
+        for (var mat of _tempMaterialSet) {
```
Comment (Maksims) [inline src/scene/forward-renderer.js:2330]: Same as above comment, regarding two loops.

File: src/scene/forward-renderer.js (hunk 1 above)
Comment (mvaligursky) [reply src/scene/forward-renderer.js]: fixed, thanks

File: src/scene/forward-renderer.js (hunk 3 above)
Comment (mvaligursky) [reply src/scene/forward-renderer.js]: a fine suggestion. I've done that now.

File: src/scene/forward-renderer.js (hunk 4 above)
Comment (mvaligursky) [reply src/scene/forward-renderer.js]: same as above

File: src/scene/scene.js (hunk 2 above)
Comment (willeastcott) [reply src/scene/scene.js:143]: OK, fair enough. 🚀

**Review by willeastcott — APPROVED** [review]

**Pull request merged**